### PR TITLE
Blank line between nested and function def in stub files.

### LINKED ISSUE #3861: Don't inspect the previous body for blank line rules in stub files
#2736 added a rule that preserves a blank line between attributes and methods. The implementation looks at the previous line instead of the previous statement, so it depends on the last statement in the (potentially arbitrarily nested) body if black inserts a blank line or not ([code](https://github.com/psf/black/blob/c160e4b7ce30c661ac4f2dfa5038becf1b8c5c33/src/black/lines.py#L599-L601)). #2783/#3564 fixes this for classes by always inserting a blank line, but this does not apply e.g. to function definitions. This affects e.g. typeshed, which uses `if sys.version_info > (3, x):` in stub files a lot.

I propose to change this to look at the type of the previous statement instead of the previous line.

**Preview style**

```python
import sys

if sys.version_info > (3, 7):
    class Nested1:
        assignment = 1
        def function_definition(self): ...

    def f1(self) -> str: ...

    class Nested2:
        def function_definition(self): ...
        assignment = 1

    def f2(self) -> str: ...

if sys.version_info > (3, 7):
    def nested1():
        assignment = 1
        def function_definition(self): ...

    def f1(self) -> str: ...
    def nested2():
        def function_definition(self): ...
        assignment = 1
    def f2(self) -> str: ...
``` 

**Desired style**

(or a different set of blank line rules that don't inspect the previous function body anymore)

```python
import sys

if sys.version_info > (3, 7):
    class Nested1:
        assignment = 1
        def function_definition(self): ...

    def f1(self) -> str: ...

    class Nested2:
        def function_definition(self): ...
        assignment = 1

    def f2(self) -> str: ...

if sys.version_info > (3, 7):
    def nested1():
        assignment = 1
        def function_definition(self): ...

    def f1(self) -> str: ...
    def nested2():
        def function_definition(self): ...
        assignment = 1

    def f2(self) -> str: ...
```

### PR DESCRIPTION
**Pull request by konstin**
### Description

The idea behind this change is that we stop looking into previous body to determine if there should be a blank before a function or class definition.

Input:

```python
import sys

if sys.version_info > (3, 7):
    class Nested1:
        assignment = 1
        def function_definition(self): ...
    def f1(self) -> str: ...
    class Nested2:
        def function_definition(self): ...
        assignment = 1
    def f2(self) -> str: ...

if sys.version_info > (3, 7):
    def nested1():
        assignment = 1
        def function_definition(self): ...
    def f1(self) -> str: ...
    def nested2():
        def function_definition(self): ...
        assignment = 1
    def f2(self) -> str: ...
```

Stable style
```python
import sys

if sys.version_info > (3, 7):
    class Nested1:
        assignment = 1
        def function_definition(self): ...

    def f1(self) -> str: ...

    class Nested2:
        def function_definition(self): ...
        assignment = 1
    def f2(self) -> str: ...

if sys.version_info > (3, 7):
    def nested1():
        assignment = 1
        def function_definition(self): ...

    def f1(self) -> str: ...
    def nested2():
        def function_definition(self): ...
        assignment = 1
    def f2(self) -> str: ...
```

In the stable formatting, we have a blank line sometimes, not depending on the previous statement on the same level, but on the last (potentially nested) statement in the previous body.

#2783/#3564 fixes this for classes in preview style:

```python
import sys

if sys.version_info > (3, 7):
    class Nested1:
        assignment = 1
        def function_definition(self): ...

    def f1(self) -> str: ...

    class Nested2:
        def function_definition(self): ...
        assignment = 1

    def f2(self) -> str: ...

if sys.version_info > (3, 7):
    def nested1():
        assignment = 1
        def function_definition(self): ...

    def f1(self) -> str: ...
    def nested2():
        def function_definition(self): ...
        assignment = 1
    def f2(self) -> str: ...
```

This PR additionally fixes this for function definitions:

```python
if sys.version_info > (3, 7):
    if sys.platform == "win32":
        assignment = 1
        def function_definition(self): ...

    def f1(self) -> str: ...
    if sys.platform != "win32":
        def function_definition(self): ...
        assignment = 1

    def f2(self) -> str: ...

if sys.version_info > (3, 8):
    if sys.platform == "win32":
        assignment = 1
        def function_definition(self): ...

    class F1: ...
    if sys.platform != "win32":
        def function_definition(self): ...
        assignment = 1

    class F2: ...
```

You can see the effect of this change on typeshed in https://github.com/konstin/typeshed/pull/1. As baseline, the preview mode changes without this PR are at https://github.com/konstin/typeshed/pull/2.

Closes #3861

### Checklist - did you ...

- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Thanks for the PR and for running this on typeshed! I'd like to hear what other typeshed maintainers (@srittau, @Akuli, @AlexWaygood, @hauntsaninja) think of the changes in https://github.com/konstin/typeshed/pull/1/files. Personally I'm supportive; they do make the code more clear.

**Comment by AlexWaygood** [comment]
> I'd like to hear what other typeshed maintainers (@srittau, @Akuli, @AlexWaygood, @hauntsaninja) think of the changes in https://github.com/konstin/typeshed/pull/1/files. Personally I'm supportive; they do make the code more clear.

Yeah, agreed, those pretty much all look like clear improvements to me! Thanks for working on this @konstin :D

File: src/black/lines.py [hunk 1]
```diff
@@ -709,6 +709,17 @@ def _maybe_empty_lines_for_class_or_def(  # noqa: C901
...
                 else:
                     newlines = 1
+            # Remove case `self.previous_line.depth > current_line.depth` below when
+            # this becomes stable.
+            #
+            # Don't inspect the previous line if it's part of the body of the previous
+            # statement in the same level, we always want a blank line if there's
+            # something with a body preceding.
+            elif (
+                Preview.blank_line_between_nested_and_def_stub_file
```
Comment (JelleZijlstra) [inline src/black/lines.py:719]:
```suggestion
                Preview.blank_line_between_nested_and_def_stub_file in current_line.mode
```

**Comment by srittau** [comment]
LGTM

**Review by JelleZijlstra — APPROVED** [review]
Looks good from an implementation perspective, will wait a little longer for feedback on whether the formatting change is good.

**Review by hauntsaninja — APPROVED** [review]

**Pull request merged**